qiskit_experiments/calibration/experiments/ef_rabi.py

```diff
@@ -0,0 +1,106 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Rabi amplitude experiment for the e-f transition."""
+
+import warnings
+from typing import Optional
+import numpy as np
+
+from qiskit.providers import Backend
+from qiskit.providers.options import Options
+from qiskit import QuantumCircuit
+from qiskit.circuit import Gate, Parameter
+import qiskit.pulse as pulse
+
+from qiskit_experiments.calibration.experiments import Rabi
+
+
+class EFRabi(Rabi):
+    """An experiment that scans the amplitude of a pulse to calibrate rotations between 1 and 2.
+
+    This experiment is similar to the Rabi experiment between the ground and first excited state.
+    The difference is in the initial X gate used to populate the first excited state. The Rabi pulse
+    is then applied on the 1 <-> 2 transition (sometimes also labeled the e <-> f transition) which
+    implies that frequency shift instructions are used. The necessary frequency shift (typically the
+    qubit anharmonicity) should be specified through the experiment options.
+
+    The circuits are of the form:
+
+    .. parsed-literal::
+
+                   ┌───┐┌───────────┐ ░ ┌─┐
+              q_0: ┤ X ├┤ Rabi(amp) ├─░─┤M├
+                   └───┘└───────────┘ ░ └╥┘
+        measure: 1/══════════════════════╩═
+                                         0
+    """
+
+    @classmethod
+    def _default_experiment_options(cls) -> Options:
+        """Default values for the pulse if no schedule is given.
+
+        Users can set a schedule by doing
+
+        .. code-block::
+
+            ef_rabi.set_experiment_options(schedule=rabi_schedule)
+
+        """
+        return Options(
+            duration=160,
+            sigma=40,
+            amplitudes=np.linspace(-0.95, 0.95, 51),
+            schedule=None,
+            normalization=True,
+            frequency_shift=0.0,
+        )
+
+    def _default_gate_schedule(self, backend: Optional[Backend] = None):
+        """Create the default schedule for the EFRabi gate with a frequency shift to the 1-2
+        transition."""
+        amp = Parameter("amp")
+        with pulse.build(backend=backend, name=self.__rabi_gate_name__) as default_schedule:
+
+            if self.experiment_options.frequency_shift == 0.0:
```

Comment:
This is nice warning! Perhaps,
```suggestion
            if np.isclose(self.experiment_options.frequency_shift, 0.):
```

However, this means we cannot stop experiment even if we notice the wrong experiment setting thanks to this warning because job is immediately triggered. Instead, I more prefer overriding `self.circuits` method with
```python
if self.experiment_options.frequency_shift is None:
   self.set_experiment_options(frequency_shift=backend.defaults["qubit_freq_est"][self.physical_qubits[0]])
return super().circuits(backend)
```
and set `frequency_shift` defaults to `None`. This code automatically populates missing frequency shift from the backend information unless user overrides it. This will save us from writing multiple lines of codes before running this experiment.

Reply:
@nkanazawa1989 what does this bring? `frequency_shift=backend.defaults()["qubit_freq_est"][self.physical_qubits[0]]` looks like you want to **shift** by the qubit frequency so if you start with 5.5GHz you end up with 11GHz.

Reply:
Sorry that's my mistake. Thanks for finding this. This is the correct code.
```python
properties = backend.properties()
anharm, _ = properties .qubit_property(self.physical_qubits[0])["anharmonicity"]

self.set_experiment_options(frequency_shift=anharm)
```

Reply:
Of course you can catch `KeyError` and raise warning or exception to ask one to set `frequency_shift`.

Reply:
Ok so since the backend provides this we can set the default frequency shift to `None`. Then we use the user provided value and if the the user does not provide one, we look for it in the backend. Finally, if the backend does not provide one we raise an error. I.e. catch `KeyError` and raise a meaningful error message from it.

Reply:
Thank you both for the very nice suggestions! These are implemented here 5d771d8.
Instead of overriding `self.circuits` in `EFRabi`, I suggest putting the query for the anharmonicity in the `self._default_gate_schedule`. That way, `self.circuits` checks first for a user-defined custom schedule and only if this is not provided, the `self._default_gate_schedule` is called from which the `frequency_shift` check from the backend is performed.